File: examples/smart_contracts/v2/python/dryrun_debugging.py
import json
import base64

from algosdk.v2client import algod
from algosdk import account, mnemonic
from algosdk.v2client.models import DryrunRequest, DryrunSource
from algosdk.future.transaction import PaymentTxn, LogicSig, LogicSigTransaction
# from algosdk.testing import dryrun

def wait_for_confirmation(client, txid):
    """
    Utility function to wait until the transaction is
    confirmed before proceeding.
    """
    last_round = client.status().get('last-round')
    txinfo = client.pending_transaction_info(txid)
    while not (txinfo.get('confirmed-round') and txinfo.get('confirmed-round') > 0):
        print("Waiting for confirmation")
        last_round += 1
        client.status_after_block(last_round)
        txinfo = client.pending_transaction_info(txid)
    print("Transaction {} confirmed in round {}.".format(
        txid, txinfo.get('confirmed-round')))
    return txinfo

# dryrun source if provided, else dryrun compiled
def dryrun_debug(lstx, mysource):
    sources = []
    if (mysource != None):
        # source
        sources = [DryrunSource(field_name="lsig", source=mysource, txn_index=0)]
    drr = DryrunRequest(txns=[lstx], sources=sources)
    dryrun_response = algod_client.dryrun(drr)
    return dryrun_response

try:
    # Create an algod client
    algod_token = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    algod_address = "http://localhost:4001"

    receiver = "ATTR6RUEHHBHXKUHT4GUOYWNBVDV2GJ5FHUWCSFZLHD55EVKZWOWSM7ABQ"
    algod_client = algod.AlgodClient(algod_token, algod_address)

    myprogram = "samplearg.teal"

    # Read TEAL program
    data = open(myprogram, 'r').read()

    # Compile TEAL program
    response = algod_client.compile(data)
    # Print(response)
    print("Response Result = ", response['result'])
    print("Response Hash = ", response['hash'])

    # Create logic sig
    programstr = response['result']
    t = programstr.encode()
    program = base64.decodebytes(t)
    print(program)
    
    # Create arg to pass
    # string parameter
    # arg_str = "<my string>"
    # arg1 = arg_str.encode()
    # lsig = transaction.LogicSig(program, args=[arg1])

    # integer parameter
    # arg1 = (123).to_bytes(8, 'big')
    # lsig = transaction.LogicSig(program, args=[arg1])
    # see more info here: https://developer.algorand.org/docs/features/asc1/sdks/#accessing-teal-program-from-sdks

    # if TEAL program requires an arg,
    # if not, omit args param on LogicSig
    # lsig = LogicSig(program)
    arg1 = (123).to_bytes(8, 'big')
    lsig = LogicSig(program, args=[arg1])

    # Recover the account that is wanting to delegate signature
    passphrase = "<25-word-mnemonic>"
    sk = mnemonic.to_private_key(passphrase)
    addr = account.address_from_private_key(sk)
    print("Address of Sender/Delegator: " + addr)

    # Sign the logic signature with an account sk
    lsig.sign(sk)

    # Get suggested parameters
    params = algod_client.suggested_params()
    # Comment out the next two (2) lines to use suggested fees
    params.flat_fee = True
    params.fee = 1000

    # Build transaction
    amount = 10000
    closeremainderto = None

    # Create a transaction
    txn = PaymentTxn(
        addr, params, receiver, amount, closeremainderto)
    # Create the LogicSigTransaction with contract account LogicSig
    lstx = LogicSigTransaction(txn, lsig)
    # transaction.write_to_file([lstx], "./simple.stxn")

    # compile
    dryrun_response_compiled = dryrun_debug(lstx, None)
    print ("COMPILED Dryrun results...")
    print(json.dumps(dryrun_response_compiled, indent=2))
   
    # source   
    dryrun_respone_source = dryrun_debug(lstx, source)
    print("SOURCE Dryrun results...")
    print(json.dumps(dryrun_respone_source, indent=2))
   
    # Send raw LogicSigTransaction to network

    txid = algod_client.send_transaction(lstx)
    print("Transaction ID: " + txid)
    wait_for_confirmation(algod_client, txid)
except Exception as e:
    print(e)

# output should look similar to this
# COMPILED Dryrun results...
# {
#     "error": "",
#     "protocol-version": "https://github.com/algorandfoundation/specs/tree/e5f565421d720c6f75cdd186f7098495caf9101f",
#     "txns": [
#         {
#             "disassembly": [
#                 "// version 1",
#                 "intcblock 123",
#                 "arg_0",
#                 "btoi",
#                 "intc_0",
#                 "==",
#                 ""
#             ],
#             "logic-sig-messages": [
#                 "PASS"
#             ],
#             "logic-sig-trace": [
#                 {
#                     "line": 1,
#                     "pc": 1,
#                     "stack": []
#                 },
#                 {
#                     "line": 2,
#                     "pc": 4,
#                     "stack": []
#                 },
#                 {
#                     "line": 3,
#                     "pc": 5,
#                     "stack": [
#                         {
#                             "bytes": "AAAAAAAAAHs=",
#                             "type": 1,
#                             "uint": 0
#                         }
#                     ]
#                 },
#                 {
#                     "line": 4,
#                     "pc": 6,
#                     "stack": [
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 123
#                         }
#                     ]
#                 },
#                 {
#                     "line": 5,
#                     "pc": 7,
#                     "stack": [
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 123
#                         },
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 123
#                         }
#                     ]
#                 },
#                 {
#                     "line": 6,
#                     "pc": 8,
#                     "stack": [
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 1
#                         }
#                     ]
#                 }
#             ]
#         }
#     ]
# }
# SOURCE Dryrun results...
# {
#     "error": "",
#     "protocol-version": "https://github.com/algorandfoundation/specs/tree/e5f565421d720c6f75cdd186f7098495caf9101f",
#     "txns": [
#         {
#             "disassembly": [
#                 "// version 1",
#                 "intcblock 123",
#                 "arg_0",
#                 "btoi",
#                 "intc_0",
#                 "==",
#                 ""
#             ],
#             "logic-sig-messages": [
#                 "PASS"
#             ],
#             "logic-sig-trace": [
#                 {
#                     "line": 1,
#                     "pc": 1,
#                     "stack": []
#                 },
#                 {
#                     "line": 2,
#                     "pc": 4,
#                     "stack": []
#                 },
#                 {
#                     "line": 3,
#                     "pc": 5,
#                     "stack": [
#                         {
#                             "bytes": "AAAAAAAAAHs=",
#                             "type": 1,
#                             "uint": 0
#                         }
#                     ]
#                 },
#                 {
#                     "line": 4,
#                     "pc": 6,
#                     "stack": [
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 123
#                         }
#                     ]
#                 },
#                 {
#                     "line": 5,
#                     "pc": 7,
#                     "stack": [
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 123
#                         },
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 123
#                         }
#                     ]
#                 },
#                 {
#                     "line": 6,
#                     "pc": 8,
#                     "stack": [
#                         {
#                             "bytes": "",
#                             "type": 2,
#                             "uint": 1
#                         }
#                     ]
#                 }
#             ]
#         }
#     ]
# }

File: examples/smart_contracts/v2/go/dryrunDebugging.go
package main

import (
	"context"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"io/ioutil"
	"log"
	"os"

	"fmt"
	"github.com/algorand/go-algorand-sdk/future"

	"github.com/algorand/go-algorand-sdk/client/v2/algod"
	modelsV2 "github.com/algorand/go-algorand-sdk/client/v2/common/models"
	"github.com/algorand/go-algorand-sdk/crypto"
	"github.com/algorand/go-algorand-sdk/mnemonic"
	"github.com/algorand/go-algorand-sdk/transaction"
	"github.com/algorand/go-algorand-sdk/types"
)



// prettyPrint prints Go structs
func prettyPrint(data interface{}) {
    var p []byte
    //    var err := error
    p, err := json.MarshalIndent(data, "", "\t")
    if err != nil {
        fmt.Println(err)
        return
    }
    fmt.Printf("%s \n", p)
}

// dryrunDebugging returns a response with disassembly, logic-sig-messages w PASS/REJECT and sig-trace
func dryrunDebugging(lsig types.LogicSig, args [][]byte,tealFile []byte, client *algod.Client) modelsV2.DryrunResponse {
    // if tealFile is nil, use compile
	txns := []types.SignedTxn{{}}
    sources := []modelsV2.DryrunSource{}
    if (tealFile == nil){
        // complile
    	txns[0].Lsig = lsig    
    }else{
        // source
        txns[0].Lsig.Args = args
		sources = append(sources, modelsV2.DryrunSource{
			FieldName: "lsig",
			Source:    string(tealFile),
            TxnIndex:  0,
        })
    }

	ddr := modelsV2.DryrunRequest{
		Txns:    txns,
		Sources: sources,
	}

	result, err := client.TealDryrun(ddr).Do(context.Background())
	if err != nil {
		return result
    }
    return  result
}

func main() {
    // sandbox
    const algodAddress = "http://localhost:4001"
    const algodToken = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    // const algodToken = "<algod-token>"
    // const algodAddress = "<algod-address>"
    // Create an algod client
    algodClient, err := algod.MakeClient(algodAddress, algodToken)
    if err != nil {
        fmt.Printf("failed to make algod client: %s\n", err)
        return
    }	
    // Get private key for sender address - 
    // mnemonics should not be used in production code. Demo purposes only
    PASSPHRASE := "<25-word-mnemonic>"
    sk, err := mnemonic.ToPrivateKey(PASSPHRASE)	
    pk := sk.Public()
    var a types.Address
    cpk := pk.(ed25519.PublicKey)
    copy(a[:], cpk[:])
    fmt.Printf("Address: %s\n", a.String())	
    sender := a.String()

    // Create logic signature
    var ma crypto.MultisigAccount

    // samplearg.teal  ... This code is meant for learning purposes only
    // It should not be used in production
    // arg_0
    // btoi
    // int 123
    // == 

    file, err := os.Open("./samplearg.teal")
    if err != nil {
        log.Fatal(err)
    }

    defer file.Close()
    tealFile, err := ioutil.ReadFile("./samplearg.teal")
    if err != nil {
        fmt.Printf("failed to read file: %s\n", err)
        return}

    response, err := algodClient.TealCompile(tealFile).Do(context.Background())
    fmt.Printf("Hash = %s\n", response.Hash)
    fmt.Printf("Result = %s\n", response.Result)
    
    program, err :=  base64.StdEncoding.DecodeString(response.Result)	
    // if no args use these two lines
    // var args [][]byte
    // lsig, err := crypto.MakeLogicSig(program, args, sk, m a)

    // string parameter
    // args := make([][]byte, 1)
    // args[0] = []byte("<my string>")
    // lsig, err := crypto.MakeLogicSig(program, args, sk, ma)
    
    // integer args parameter
    args := make([][]byte, 1)
    var buf [8]byte
    binary.BigEndian.PutUint64(buf[:], 123)
    args[0] = buf[:]
    lsig, err := crypto.MakeLogicSig(program, args, sk, ma)
    

    // tealDryrunResult.response = result
    
    // Construct the transaction
    txParams, err := algodClient.SuggestedParams().Do(context.Background())
    if err != nil {
        fmt.Printf("Error getting suggested tx params: %s\n", err)
        return
    }
    // comment out the next two (2) lines to use suggested fees
    // txParams.FlatFee = true
    // txParams.Fee = 1000

    // Make transaction
    // const receiver = "<receiver-address>"
    const receiver = "QUDVUXBX4Q3Y2H5K2AG3QWEOMY374WO62YNJFFGUTMOJ7FB74CMBKY6LPQ"	
    // const fee = <fee>
    // const amount = <amount>
    var fee = uint64(txParams.MinFee)
    const amount = 100000
    note := []byte("Hello World")
    genID := txParams.GenesisID
    genHash := txParams.GenesisHash
    firstValidRound := uint64(txParams.FirstRoundValid)
    lastValidRound := uint64(txParams.LastRoundValid)
    tx, err := transaction.MakePaymentTxnWithFlatFee(
        sender, receiver, fee, amount, firstValidRound, lastValidRound,
        note, "", genID, genHash )

    txID, stx, err := crypto.SignLogicSigTransaction(lsig, tx)
    if err != nil {
        fmt.Printf("Signing failed with %v", err)
        return
    }
    fmt.Printf("Signed tx: %v\n", txID)
    // logic signature transaction can be written to a file
    // f, err := os.Create("simple.stxn")

    // defer f.Close()
    // if _, err := f.Write(stx); err != nil {
    //     // handle
    // }
    // if err := f.Sync(); err != nil {
    //     // handle
    // }

    // Submit the raw transaction as normal 
    // compile
    result := dryrunDebugging(lsig, args, nil, algodClient)

    fmt.Printf("Compile Result = ")
    prettyPrint(result)
    // source
    result = dryrunDebugging(lsig, args, tealFile, algodClient)
    fmt.Printf("Source Result = ")
    prettyPrint(result)

    transactionID, err := algodClient.SendRawTransaction(stx).Do(context.Background())
    if err != nil {
        fmt.Printf("Sending failed with %v\n", err)
    }
    fmt.Printf("Transaction ID: %v\n", transactionID)

    // Wait for confirmation
	confirmedTxn, err := transaction.WaitForConfirmation(algodClient,txID,  4, context.Background())
	if err != nil {
		fmt.Printf("Error waiting for confirmation on txID: %s\n", txID)
		return
	}
	fmt.Printf("Confirmed Transaction: %s in Round %d\n", txID ,confirmedTxn.ConfirmedRound)



    }
// the response should look similar to this...
// Result = {
// 	"protocol-version": "https://github.com/algorandfoundation/specs/tree/e5f565421d720c6f75cdd186f7098495caf9101f",
// 	"txns": [
// 		{
// 			"disassembly": [
// 				"// version 1",
// 				"intcblock 123",
// 				"arg_0",
// 				"btoi",
// 				"intc_0",
// 				"==",
// 				""
// 			],
// 			"logic-sig-messages": [
// 				"PASS"
// 			],
// 			"logic-sig-trace": [
// 				{
// 					"line": 1,
// 					"pc": 1
// 				},
// 				{
// 					"line": 2,
// 					"pc": 4
// 				},
// 				{
// 					"line": 3,
// 					"pc": 5,
// 					"stack": [
// 						{
// 							"bytes": "AAAAAAAAAHs=",
// 							"type": 1
// 						}
// 					]
// 				},
// 				{
// 					"line": 4,
// 					"pc": 6,
// 					"stack": [
// 						{
// 							"type": 2,
// 							"uint": 123
// 						}
// 					]
// 				},
// 				{
// 					"line": 5,
// 					"pc": 7,
// 					"stack": [
// 						{
// 							"type": 2,
// 							"uint": 123
// 						},
// 						{
// 							"type": 2,
// 							"uint": 123
// 						}
// 					]
// 				},
// 				{
// 					"line": 6,
// 					"pc": 8,
// 					"stack": [
// 						{
// 							"type": 2,
// 							"uint": 1
// 						}
// 					]
// 				}
// 			]
// 		}

// 	]
// } 